nx/lib/nx/defn/evaluator.ex

```diff
@@ -469,4 +477,51 @@ defmodule Nx.Defn.Evaluator do
   defp composite_to_params(other, acc) do
     [fn -> other end | acc]
   end
+
+  # Debug hook: print/save node info if enabled and not already printed
+  defp maybe_debug_node(ans, res, %{debug_options: nil} = state), do: {res, state}
+  defp maybe_debug_node(%Nx.Tensor{data: %Expr{id: id, op: op, args: args}} = ans, res, %{debug_options: opts, printed_nodes: printed_nodes} = state) do
+    if MapSet.member?(printed_nodes, id) do
+      {res, state}
+    else
+      inspect_limit = Keyword.get(opts, :inspect_limit, 10)
+      save_path = Keyword.get(opts, :save_path, nil)
+      node_info = format_node_info(ans, args, res, inspect_limit)
+      if save_path do
+        :ok = save_node_info_to_file(save_path, id, node_info)
+      else
+        IO.puts(node_info)
+      end
+      {res, %{state | printed_nodes: MapSet.put(printed_nodes, id)}}
+    end
+  end
+
+  defp format_node_info(%Nx.Tensor{data: %Expr{id: id, op: op, args: args}}, args_list, res, inspect_limit) do
+    args_str =
+      Enum.map(args_list, fn arg ->
+        case arg do
+          %Nx.Tensor{data: %Expr{id: arg_id}} ->
```

Comment:
I wonder if we can just add an option to defn expr inspect to print the id... Then we'd just call inspect on the map for all args